### PR TITLE
[imitation_game] added an android test

### PR DESCRIPTION
**Pull request by gaaclarke**
Added a test that shows how to write tests for android.  This test uses adb to measure the memory usage of the smiley app.  This is also an example of how to do measurements from the host computer instead of the device.

I also had to make the results get saved to disk and edited since we can only run a subset of the tests in one run (iOS or Android).

### CODE REVIEW

File: packages/imitation_game/bin/imitation_game.dart [hunk 1]
```diff
@@ -36,13 +37,39 @@ Future<List<FileSystemEntity>> findFiles(Directory dir, {FileFilter where}) {
...
+  newResults.forEach((String test, dynamic testValue) {
+    // ignore: avoid_as
+    final Map<String, dynamic> testMap = testValue as Map<String, dynamic>;
+    testMap.forEach((String platform, dynamic platformValue) {
+      // ignore: avoid_as
+      final Map<String, dynamic> platformMap =
+          // ignore: avoid_as
+          platformValue as Map<String, dynamic>;
+      platformMap.forEach((String measurement, dynamic measurementValue) {
+        if (!result.containsKey(test)) {
```
Comment (liyuqian) [inline packages/imitation_game/bin/imitation_game.dart:60]: Nit: `result.putIfAbsent(...)`

File: packages/imitation_game/README.md [hunk 2]
```diff
@@ -93,11 +93,14 @@ An implementation has to follow these rules:
 A single test run can report multiple numbers.
 
 ## Results
-Date created: 2020-08-17 23:57:16.702500Z
+Date created: 2020-08-19 20:46:17.734351Z
 
 - smiley
     - flutter
-      - startupTime: 0.561475s
+      - startupTime: 0.691717
```
Comment (liyuqian) [inline packages/imitation_game/README.md:100]: As mentioned in [another comment][1], we may need to add an OS subdirectory as `smiley/android/flutter` and `smiley/ios/flutter`. Otherwise, it will be confusing if the `startupTime` is from iOS or Android as they're different. For example, the `startupTime` here is from iOS while the `adb_memory_total` is from Android. However, it's very easy to think that `startupTime` here is also coming from Android.

[1]: https://github.com/flutter/packages/pull/181#discussion_r460359393

File: packages/imitation_game/imitation_tests/smiley/android/app/src/test/java/dev/flutter/imitation_game/smiley/ExampleUnitTest.kt [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package dev.flutter.imitation_game.smiley
```
Comment (liyuqian) [inline packages/imitation_game/imitation_tests/smiley/android/app/src/test/java/dev/flutter/imitation_game/smiley/ExampleUnitTest.kt:1]: Is this file necessary?

File: packages/imitation_game/imitation_tests/smiley/android/run_android.sh [hunk 4]
```diff
@@ -0,0 +1,14 @@
+set -e
+cd $( dirname "${BASH_SOURCE[0]}" )
+source ../../../helper/run_android_helper.sh
+
+readonly BUNDLE="dev.flutter.imitation_game.smiley"
+readonly TEST_NAME="smiley"
+
+# TODO(gaaclarke): Get assembleRelease working.  I don't know if the performance
+# is any different but assembleRelease requires signing the apk.
```
Comment (xster) [inline packages/imitation_game/imitation_tests/smiley/android/run_android.sh:9]: No it doesn't https://github.com/flutter/flutter/blob/eef4220a878fd5dfe1551f76a78168347d6aa9eb/examples/hello_world/android/app/build.gradle#L47

File: packages/imitation_game/imitation_tests/smiley/android/run_android.sh [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+./gradlew assembleDebug
+launch_apk ./app/build/outputs/apk/debug/app-debug.apk $BUNDLE "MainActivity"
+sleep 10
```
Comment (xster) [inline packages/imitation_game/imitation_tests/smiley/android/run_android.sh:12]: you can probably stream adb logcat for a success/failure result. @blasten might have other ideas too.

File: packages/imitation_game/run.sh [hunk 6]
```diff
@@ -1 +1 @@
-pub run imitation_game
+pub run imitation_game --platform=android
```
Comment (xster) [inline packages/imitation_game/run.sh:1]: you want to hardcode this?

File: packages/imitation_game/last_results.json [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+  "smiley": {
+    "flutter": {
+      "startupTime": 0.691717,
+      "adb_memory_total": 101928.0
+    },
+    "uikit": {
+      "startupTime": 0.2632870674133301
+    },
+    "android": {
+      "adb_memory_total": 35267.0
```
Comment (xster) [inline packages/imitation_game/last_results.json:11]: how come Android doesn't have a startup time?

File: packages/imitation_game/last_results.json [hunk 8]
```diff
@@ -0,0 +1,14 @@
+{
+  "smiley": {
+    "flutter": {
+      "startupTime": 0.691717,
+      "adb_memory_total": 101928.0
```
Comment (xster) [inline packages/imitation_game/last_results.json:5]: man this is brutal. Where's the code that populated this measurement? I assume this is a strict delta measurement between before Flutter started and after Flutter started? Did we call the gc before each measurement? Can we break it down into memory categories to make this more actionable?

File: packages/imitation_game/README.md (hunk 2 above)
Comment (gaaclarke) [reply packages/imitation_game/README.md:100]: Yea, I thought about that.  I think we can just disambiguate the measurements with names like "ios_startup_time".  I changed that measurement's name.

File: packages/imitation_game/bin/imitation_game.dart (hunk 1 above)
Comment (gaaclarke) [reply packages/imitation_game/bin/imitation_game.dart:60]: done

File: packages/imitation_game/imitation_tests/smiley/android/app/src/test/java/dev/flutter/imitation_game/smiley/ExampleUnitTest.kt (hunk 3 above)
Comment (gaaclarke) [reply packages/imitation_game/imitation_tests/smiley/android/app/src/test/java/dev/flutter/imitation_game/smiley/ExampleUnitTest.kt]: Nope, deleted.

File: packages/imitation_game/last_results.json (hunk 8 above)
Comment (gaaclarke) [reply packages/imitation_game/last_results.json:5]: It's just the smiley application.  If you run `adb shell dumpsys meminfo $BUNDLE_NAME` after launching that app on an android device you can see the breakdown.  Figuring out how to get the release build working may help a bit.

File: packages/imitation_game/last_results.json (hunk 7 above)
Comment (gaaclarke) [reply packages/imitation_game/last_results.json:11]: No one implemented it.

File: packages/imitation_game/last_results.json (hunk 8 above)
Comment (gaaclarke) [reply packages/imitation_game/last_results.json:5]: Also, I don't think it would be fair to call "GC" before measuring since that isn't how people are suppose to write their apps.

Comment (gaaclarke) [reply packages/imitation_game/last_results.json]: I ran this again today and with today's flutter, flutter wins.  I'll add the flutter version to the results...

File: packages/imitation_game/run.sh (hunk 6 above)
Comment (gaaclarke) [reply packages/imitation_game/run.sh]: added an argument

File: packages/imitation_game/bin/imitation_game.dart [hunk 9]
```diff
@@ -36,13 +37,39 @@ Future<List<FileSystemEntity>> findFiles(Directory dir, {FileFilter where}) {
...
   values['date'] = DateTime.now().toUtc();
   final String output = template.renderString(values);
   return output;
 }
 
+Map<String, dynamic> _integrate(
+    {Map<String, dynamic> oldResults, Map<String, dynamic> newResults}) {
+  final Map<String, dynamic> result = Map<String, dynamic>.from(oldResults);
+  newResults.forEach((String test, dynamic testValue) {
+    // ignore: avoid_as
```
[the remark this answers is not in the document]
Comment (gaaclarke) [reply packages/imitation_game/bin/imitation_game.dart]: I threw in a file ignore.

File: packages/imitation_game/bin/imitation_game.dart [hunk 10]
```diff
@@ -36,13 +37,39 @@ Future<List<FileSystemEntity>> findFiles(Directory dir, {FileFilter where}) {
...
 String _makeMarkdownOutput(Map<String, dynamic> results) {
+  // TODO(gaaclarke): Add the Flutter version.
   final Template template = Template(readmeTemplate, name: 'README.md');
   final Map<String, dynamic> values = Map<String, dynamic>.from(results);
   values['date'] = DateTime.now().toUtc();
   final String output = template.renderString(values);
   return output;
 }
 
+Map<String, dynamic> _integrate(
```
[the remark this answers is not in the document]
Comment (gaaclarke) [reply packages/imitation_game/bin/imitation_game.dart:53]: Done.

File: packages/imitation_game/last_results.json (hunk 8 above)
Comment (gaaclarke) [reply packages/imitation_game/last_results.json]: (Done, added the flutter version in the output)

Comment (xster) [reply packages/imitation_game/last_results.json]: ah good to know you're getting it just for the process. I think GC before measuring is fairly standard. It's not fair if you only do it in Flutter. I'd expect GC to be called for both the Android and Flutter trials. Otherwise it'd be "non-hermetic" and repeat runs won't be consistent.

File: packages/imitation_game/README.md [hunk 11]
```diff
@@ -93,11 +93,18 @@ An implementation has to follow these rules:
...
+Flutter 1.22.0-10.0.pre.82 • channel master • https:&#x2F;&#x2F;github.com&#x2F;flutter&#x2F;flutter.git
+Framework • revision 9248fda410 (35 minutes ago) • 2020-09-03 16:08:01 -0700
+Engine • revision 3f05b521b9
+Tools • Dart 2.10.0 (build 2.10.0-86.0.dev)
 
 - smiley
     - flutter
-      - startupTime: 0.561475s
+      - ios_startup_time: 0.691717
+      - adb_memory_total: 45148.0
```
Comment (xster) [inline packages/imitation_game/README.md:105]: Are we sure this is right? Seems a bit surprising... Maybe a quick sanity check in Android Studio's memory profiler? :D

**Comment by xster** [comment]
LGTM

File: packages/imitation_game/imitation_tests/smiley/android/run_android.sh (hunk 4 above)
Comment (gaaclarke) [reply packages/imitation_game/imitation_tests/smiley/android/run_android.sh:9]: If I use assembleRelase I get this error:
```
aaclarke-macbookpro2:android aaclarke$ adb install ./app/build/outputs/apk/release/app-release-unsigned.apk
Performing Streamed Install
adb: failed to install ./app/build/outputs/apk/release/app-release-unsigned.apk: Failure [INSTALL_PARSE_FAILED_NO_CERTIFICATES: Package /data/app/vmdl1987488777.tmp/base.apk has no certificates at entry AndroidManifest.xml]
```

Googling that error leads to a myriad of suggestions.

Comment (gaaclarke) [reply packages/imitation_game/imitation_tests/smiley/android/run_android.sh:9]: Got help via chat, thanks fixed it, works for assembleRelease now.

File: packages/imitation_game/imitation_tests/smiley/android/run_android.sh [hunk 12]
```diff
@@ -7,8 +7,8 @@ readonly TEST_NAME="smiley"
 
 # TODO(gaaclarke): Get assembleRelease working.  I don't know if the performance
```
Comment (xster) [inline packages/imitation_game/imitation_tests/smiley/android/run_android.sh:8]: nit: You can remove this text now :)

Comment (gaaclarke) [reply packages/imitation_game/imitation_tests/smiley/android/run_android.sh]: done
